README.md
EngFormat-Cpp
=============

C++ based Engineering Notation Formatter

The C++ engineering formatter presented here is partly based on [code found on stackoverflow](http://stackoverflow.com/a/15734251/437272) [1] and partly based on the code by Jukka Korpela and David Hoerl as available from the project [EngineeringNotationFormatter](https://github.com/dhoerl/EngineeringNotationFormatter) [2] on GitHub.

Example usage
-------------
```Cpp
#include "eng_format.hpp"
#include <iostream>

int main()
{
    std::cout << to_engineering_string( 1234, 3, eng_prefixed, "Pa") << std::endl;
}
```

Compile and run:
```
prompt>g++ -Wall -Wextra -I../src -o example1.exe example1.cpp ../src/eng_format.cpp && example1
1.23 kPa
```

Basic C++ interface
-------------------
```Cpp
/**
 * convert a double to the specified number of digits in SI (prefix) or
 * exponential notation, optionally followed by a unit.
 */
std::string
to_engineering_string( double value, int digits, bool exponential, std::string unit = "", std::string separator = " " );

/**
 * convert the output of to_engineering_string() into a double.
 */
double
from_engineering_string( std::string text );

/**
 * step a value by the smallest possible increment.
 */
std::string
step_engineering_string( std::string text, int digits, bool exponential, bool increment );
```

Expanded C++ interface
----------------------

```Cpp
extern struct eng_prefixed_t {} eng_prefixed;
extern struct eng_exponential_t {} eng_exponential;

const bool eng_increment = true;
const bool eng_decrement = false;

/**
 * convert a double to the specified number of digits in SI (prefix) notation,
 * optionally followed by a unit.
 */
std::string
to_engineering_string( double value, int digits, eng_prefixed_t, std::string unit = "", std::string separator = " " );

/**
 * convert a double to the specified number of digits in exponential notation,
 * optionally followed by a unit.
 */
std::string
to_engineering_string( double value, int digits, eng_exponential_t, std::string unit = "", std::string separator = " " );

/**
 * step a value by the smallest possible increment, using SI notation.
 */
std::string
step_engineering_string( std::string text, int digits, eng_prefixed_t, bool increment );

/**
 * step a value by the smallest possible increment, using exponential notation.
 */
std::string
step_engineering_string( std::string text, int digits, eng_exponential_t, bool increment );
```

Notes and References
--------------------

[1] [scls](http://stackoverflow.com/users/1609077/scls). [Convert float number to string with engineering notation (with SI prefixe) in Python [closed]](http://stackoverflow.com/questions/15733772/convert-float-number-to-string-with-engineering-notation-with-si-prefixe-in-py). StackOverflow.

[2] David Hoerl. [EngineeringNotationFormatter](https://github.com/dhoerl/EngineeringNotationFormatter). iOS Project demoing C-based Engineering Notation Formatter.


CHANGES.md
Changes
=======

0.3.0 &ndash; 2 March 2015
----------------------------
- Adapt from_engineering_string() to accept an empty separator and variable-length SI prefixes, such as a UTF-8 encoded micro, "\xce\xbc" (Thanks to Josh Kelley)

0.2.0 &ndash; 12 February 2015
----------------------------
- Add separator parameter to to_engineering_string()
- Failing test: test_eng_format.cpp:115: from_engineering_string() needs adaptation 
  to support empty separator, e.g. "1k"

0.1.0 &ndash; 1 July 2013
-----------------------
- Start version numbering
- Failing test: test_eng_format.cpp:145: formats as '-1000.000e-27', should be "-1.00000 y"
- Failing test: test_eng_format.cpp:158: formats as '100.0 z', should be "100 z"


test/test_eng_format.cpp
// Copyright (C) 2013 by Martin Moene
// Copyright (C) 2013 by David Hoerl
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "lest.hpp"
#include "eng_format.hpp"

#include <cmath>
#include <iostream>
#include <limits>
#include <string>

std::string to_string( std::string  const & text ) { return text; };
std::string to_string( char const * const   text ) { return text; };

bool approx( double const a, double const b )
{
#if 0
    const double scale = 1.0;
    const double epsilon = 100 * std::numeric_limits<double>::epsilon();

    return fabs(a - b) < epsilon * (scale + (std::max)( fabs(a), fabs(b) ) );
#else
    double dbls[2] = { a, b };

    for ( int i = 0; i < 2; ++i )
    {
        int classify = std::fpclassify( dbls[i] );

        if ( classify != FP_NORMAL )
        {
            switch( classify )
            {
            case FP_SUBNORMAL:
                std::cout << "FP Error: subnormal!" << std::endl;
                return false;
            case FP_ZERO:
                break;
            case FP_INFINITE:
                std::cout << "FP Error: Infinite!" << std::endl;
                return false;
            case FP_NAN:
                std::cout << "FP Error: NaN!" << std::endl;
                return false;
            default:
                std::cout << "FP Error: unknown!" << std::endl;
                return false;
            }
        }
    }

    return ! std::isnormal( a - b );
#endif
}

const lest::test specification[] =
{
    CASE( "number converts well to string without prefix" )
    {
        EXPECT( "123" == to_engineering_string( 123, 3, false ) );
        EXPECT( "123" == to_engineering_string( 123, 3, eng_prefixed ) );
    },

    CASE( "number converts well to string using prefix" )
    {
        EXPECT( "1.23 k" == to_engineering_string( 1230, 3, false ) );
        EXPECT( "1.23 k" == to_engineering_string( 1230, 3, eng_prefixed ) );
    },

    CASE( "number converts well to string using prefix without spaces" )
    {
        EXPECT( "1.23k" == to_engineering_string( 1230, 3, false, "", "" ) );
        EXPECT( "1.23k" == to_engineering_string( 1230, 3, eng_prefixed, "", "" ) );
    },

    CASE( "number converts well to string using exponent" )
    {
        EXPECT( "1.23e3" == to_engineering_string( 1230, 3, true ) );
        EXPECT( "1.23e3" == to_engineering_string( 1230, 3, eng_exponential ) );
    },

    CASE( "number below 1000 converts well to string using exponent \"e0\"" )
    {
        EXPECT( "123e0" == to_engineering_string( 123, 3, true ) );
        EXPECT( "123e0" == to_engineering_string( 123, 3, eng_exponential ) );
    },

    CASE( "string using prefix converts well to number" )
    {
        EXPECT( lest::approx( 98.76e-3 ) == from_engineering_string( "98.76 m" ) );
        
        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76 m" ) ) );
        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76 ml" ) ) );
    },

    CASE( "string using prefix and empty separator converts well to number" )
    {
        EXPECT( lest::approx( 98.76e-3 ) == from_engineering_string( "98.76m" ) );

        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76m" ) ) );
        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76ml" ) ) );
    },

    CASE( "string using exponent converts well to number" )
    {
        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76e-3" ) ) );
        EXPECT( approx( 98.76e-3, from_engineering_string( "98.76e-3 l" ) ) );
    },

    CASE( "unit appropriately follows number converted to string" )
    {
#ifdef TEST_C_VERSION_BY_DH
        EXPECT( "1.23 kPa"  == to_engineering_string_units( 1230, 3, false, "Pa" ) );
        EXPECT( "1.23 kPa"  == to_engineering_string_units( 1230, 3, eng_prefixed, "Pa" ) );

        EXPECT( "1.23e3 Pa" == to_engineering_string_units( 1230, 3, true, "Pa"  ) );
        EXPECT( "1.23e3 Pa" == to_engineering_string_units( 1230, 3, eng_exponential, "Pa" ) );
#else
        EXPECT( "1.23 kPa"  == to_engineering_string( 1230, 3, false, "Pa" ) );
        EXPECT( "1.23 kPa"  == to_engineering_string( 1230, 3, eng_prefixed, "Pa" ) );

        EXPECT( "1.23e3 Pa" == to_engineering_string( 1230, 3, true, "Pa"  ) );
        EXPECT( "1.23e3 Pa" == to_engineering_string( 1230, 3, eng_exponential, "Pa" ) );
#endif
    },

    CASE( "zero converts well to string" )
    {
        EXPECT( "0.00"      == to_engineering_string( 0, 3, eng_prefixed ) );
        EXPECT( "0.00e0"    == to_engineering_string( 0, 3, eng_exponential ) );
    },

    CASE( "extreme converts well to string" )
    {
        EXPECT( "100e96"    == to_engineering_string( 1e98, 3, eng_exponential ) );
        EXPECT( "-10.0e-99" == to_engineering_string( -1e-98, 3, eng_exponential ) );

        EXPECT( "999.999 Y" == to_engineering_string( 999.999e24, 6, eng_prefixed ) );
        EXPECT("-1.00000 y" == to_engineering_string( -999.9999e-27, 6, eng_prefixed ) );
    },

    CASE( "prefix-exceeding value converts well to string" )
    {
        EXPECT( "100e96"    == to_engineering_string( 1e98, 3, eng_prefixed ) );
        EXPECT( "-10.0e-99" == to_engineering_string( -1e-98, 3, eng_prefixed ) );
    },

    CASE( "value rounds properly to string" )
    {
        // need the final 1 to force number past .950:
        EXPECT( "100 z"    == to_engineering_string( 99.951e-21, 3, eng_prefixed ) );
        EXPECT( "99.9 z"   == to_engineering_string( 99.949e-21, 3, eng_prefixed ) );

        EXPECT( "100.09 z" == to_engineering_string( 100.0949e-21, 5, eng_prefixed ) );
        EXPECT( "99.999 z" == to_engineering_string( 99.99851e-21, 5, eng_prefixed ) );
    },

    CASE( "bad floating point number converts accordingly to string" )
    {
        EXPECT( "NaN"      == to_engineering_string( NAN     , 3, eng_prefixed ) );
        EXPECT( "INFINITE" == to_engineering_string( INFINITY, 3, eng_exponential ) );

#pragma warning "FIX ME!"
        //EXPECT( isnan(  from_engineering_string( " " ) ) );
        //EXPECT( isnan(  from_engineering_string( "Howdie" ) ) );
        //EXPECT( isnan(  from_engineering_string( "1 Q" ) ) );
        //EXPECT( fpclassify( from_engineering_string( "0" ) ) & FP_ZERO );
    },

    CASE( "string using prefix converts well to number (y-Y)" )
    {
        EXPECT( approx( 1e-24, from_engineering_string( "1 y" ) ) );
        EXPECT( approx( 1e-21, from_engineering_string( "1 z" ) ) );
        EXPECT( approx( 1e-18, from_engineering_string( "1 a" ) ) );
        EXPECT( approx( 1e-15, from_engineering_string( "1 f" ) ) );
        EXPECT( approx( 1e-12, from_engineering_string( "1 p" ) ) );
        EXPECT( approx( 1e-9 , from_engineering_string( "1 n" ) ) );
        EXPECT( approx( 1e-6 , from_engineering_string( "1 u" ) ) );
        EXPECT( approx( 1e-3 , from_engineering_string( "1 m" ) ) );
        EXPECT( approx( 1    , from_engineering_string( "1 "  ) ) );
        EXPECT( approx( 1    , from_engineering_string( "1"   ) ) );
        EXPECT( approx( 1e+3 , from_engineering_string( "1 k" ) ) );
        EXPECT( approx( 1e+6 , from_engineering_string( "1 M" ) ) );
        EXPECT( approx( 1e+9 , from_engineering_string( "1 G" ) ) );
        EXPECT( approx( 1e+12, from_engineering_string( "1 T" ) ) );
        EXPECT( approx( 1e+15, from_engineering_string( "1 P" ) ) );
        EXPECT( approx( 1e+18, from_engineering_string( "1 E" ) ) );
        EXPECT( approx( 1e+21, from_engineering_string( "1 Z" ) ) );
        EXPECT( approx( 1e+24, from_engineering_string( "1 Y" ) ) );
    },

    CASE( "value with exponent converts well to string using prefix (easy)" )
    {
        EXPECT( "1.23 y" == to_engineering_string( 1.23e-24, 3, eng_prefixed ) );
        EXPECT( "1.23 z" == to_engineering_string( 1.23e-21, 3, eng_prefixed ) );
        EXPECT( "1.23 a" == to_engineering_string( 1.23e-18, 3, eng_prefixed ) );
        EXPECT( "1.23 f" == to_engineering_string( 1.23e-15, 3, eng_prefixed ) );
        EXPECT( "1.23 p" == to_engineering_string( 1.23e-12, 3, eng_prefixed ) );
        EXPECT( "1.23 n" == to_engineering_string( 1.23e-9 , 3, eng_prefixed ) );
        EXPECT( "1.23 u" == to_engineering_string( 1.23e-6 , 3, eng_prefixed ) );
        EXPECT( "1.23 m" == to_engineering_string( 1.23e-3 , 3, eng_prefixed ) );
        EXPECT( "1.23"   == to_engineering_string( 1.23    , 3, eng_prefixed ) );
        EXPECT( "1.23 k" == to_engineering_string( 1.23e+3 , 3, eng_prefixed ) );
        EXPECT( "1.23 M" == to_engineering_string( 1.23e+6 , 3, eng_prefixed ) );
        EXPECT( "1.23 G" == to_engineering_string( 1.23e+9 , 3, eng_prefixed ) );
        EXPECT( "1.23 T" == to_engineering_string( 1.23e+12, 3, eng_prefixed ) );
        EXPECT( "1.23 P" == to_engineering_string( 1.23e+15, 3, eng_prefixed ) );
        EXPECT( "1.23 E" == to_engineering_string( 1.23e+18, 3, eng_prefixed ) );
        EXPECT( "1.23 Z" == to_engineering_string( 1.23e+21, 3, eng_prefixed ) );
        EXPECT( "1.23 Y" == to_engineering_string( 1.23e+24, 3, eng_prefixed ) );
    },

    CASE( "value with exponent converts well to string with prefix (hard)" )
    {
        EXPECT( "1.00 y" == to_engineering_string( 1e-24, 3, eng_prefixed ) );
        EXPECT( "1.00 z" == to_engineering_string( 1e-21, 3, eng_prefixed ) );
        EXPECT( "1.00 a" == to_engineering_string( 1e-18, 3, eng_prefixed ) );
        EXPECT( "1.00 f" == to_engineering_string( 1e-15, 3, eng_prefixed ) );
        EXPECT( "1.00 p" == to_engineering_string( 1e-12, 3, eng_prefixed ) );
        EXPECT( "1.00 n" == to_engineering_string( 1e-9 , 3, eng_prefixed ) );
        EXPECT( "1.00 u" == to_engineering_string( 1e-6 , 3, eng_prefixed ) );
        EXPECT( "1.00 m" == to_engineering_string( 1e-3 , 3, eng_prefixed ) );
        EXPECT( "1.00"   == to_engineering_string( 1    , 3, eng_prefixed ) );
        EXPECT( "1.00 k" == to_engineering_string( 1e+3 , 3, eng_prefixed ) );
        EXPECT( "1.00 M" == to_engineering_string( 1e+6 , 3, eng_prefixed ) );
        EXPECT( "1.00 G" == to_engineering_string( 1e+9 , 3, eng_prefixed ) );
        EXPECT( "1.00 T" == to_engineering_string( 1e+12, 3, eng_prefixed ) );
        EXPECT( "1.00 P" == to_engineering_string( 1e+15, 3, eng_prefixed ) );
        EXPECT( "1.00 E" == to_engineering_string( 1e+18, 3, eng_prefixed ) );
        EXPECT( "1.00 Z" == to_engineering_string( 1e+21, 3, eng_prefixed ) );
        EXPECT( "1.00 Y" == to_engineering_string( 1e+24, 3, eng_prefixed ) );
    },

    CASE( "round-trip conversion on number without exponent succeeds" )
    {
        EXPECT( approx( 54.32, from_engineering_string( to_engineering_string( 54.32, 4, false ) ) ) );
        EXPECT( approx( 54.32, from_engineering_string( to_engineering_string( 54.32, 4, true  ) ) ) );
    },

    CASE( "round-trip conversion on number with exponent succeeds" )
    {
        EXPECT( approx( 543.2e2, from_engineering_string( to_engineering_string( 543.2e2, 4, false ) ) ) );
        EXPECT( approx( 543.2e2, from_engineering_string( to_engineering_string( 543.2e2, 4, true  ) ) ) );
    },

    CASE( "step up succeeds for string without prefix or exponent" )
    {
        EXPECT( "1.01"   == step_engineering_string( "1.0"   , 3, eng_prefixed, eng_increment ) );
        EXPECT( "1.01e0" == step_engineering_string( "1.0"   , 3, eng_exponential, eng_increment ) );
    },

    CASE( "step down succeeds for string without prefix or exponent" )
    {
        EXPECT( "1.00"   == step_engineering_string( "1.01"  , 3, eng_prefixed, eng_decrement ) );
        EXPECT( "1.00e0" == step_engineering_string( "1.01"  , 3, eng_exponential, eng_decrement ) );
    },

    CASE( "step up succeeds for string with zero value" )
    {
        EXPECT( "10.0 m"  == step_engineering_string( "0.0"  , 3, eng_prefixed, eng_increment ) );
        EXPECT( "10.0e-3" == step_engineering_string( "0.0"  , 3, eng_exponential, eng_increment ) );
    },

    CASE( "step up succeeds for string with prefix" )
    {
        EXPECT( "1.01 k" == step_engineering_string( "1.0 k" , 3, eng_prefixed, eng_increment ) );
        EXPECT( "1.01e3" == step_engineering_string( "1.0 k" , 3, eng_exponential, eng_increment ) );
    },

    CASE( "step down succeeds for string with prefix" )
    {
        EXPECT( "1.00 k" == step_engineering_string( "1.01 k", 3, eng_prefixed, eng_decrement ) );
        EXPECT( "1.00e3" == step_engineering_string( "1.01 k", 3, eng_exponential, eng_decrement ) );
    },

    CASE( "step up succeeds for string with exponent" )
    {
        EXPECT( "1.01 k" == step_engineering_string( "1.0e3" , 3, eng_prefixed, eng_increment ) );
        EXPECT( "1.01e3" == step_engineering_string( "1.0e3" , 3, eng_exponential, eng_increment ) );
    },

    CASE( "step down succeeds for string with exponent" )
    {
        EXPECT( "1.00 k" == step_engineering_string( "1.01e3", 3, eng_prefixed, eng_decrement ) );
        EXPECT( "1.00e3" == step_engineering_string( "1.01e3", 3, eng_exponential, eng_decrement ) );
    },

    CASE( "step up to prefix 'k' succeeds for string without prefix" )
    {
        EXPECT( "1.00 k" == step_engineering_string( "999"   , 3, eng_prefixed, eng_increment ) );
    },

    CASE( "step down to loose prefix succeeds for string with prefix 'k'" )
    {
        EXPECT( "999.0"  == step_engineering_string( "1.000 k", 4, eng_prefixed, eng_decrement ) );
    },

    CASE( "step up to loose prefix succeeds for string with prefix 'm'" )
    {
        EXPECT( "1.00"  == step_engineering_string( "999 m"  , 3, eng_prefixed, eng_increment ) );
    },

    CASE( "step down to use prefix or exponent succeeds for string without prefix" )
    {
        EXPECT( "990 m"  == step_engineering_string( "1.0"   , 3, eng_prefixed, eng_decrement ) );
        EXPECT( "990e-3" == step_engineering_string( "1.0"   , 3, eng_exponential, eng_decrement ) );
    },

    CASE( "step up to next prefix or exponent succeeds for string with prefix" )
    {
        EXPECT( "1.00 G" == step_engineering_string( "999 M" , 3, eng_prefixed, eng_increment ) );
        EXPECT( "1.00e9" == step_engineering_string( "999 M" , 3, eng_exponential, eng_increment ) );
    },

    CASE( "step down to next prefix or exponent succeeds for string with prefix" )
    {
        EXPECT( "990 k" == step_engineering_string( "1.0 M"  , 3, eng_prefixed, eng_decrement ) );
        EXPECT( "990e3" == step_engineering_string( "1.0 M"  , 3, eng_exponential, eng_decrement ) );
    },
};

int main( int argc, char* argv[] )
{
    return lest::run( specification, argc, argv );
}

// lest test framework requires C++11 lambda: VC2013 preview, g++ -std=c++11

// compilation control to test C routine to_engineering_string_units(): -DTEST_C_VERSION_BY_DH


// cl -nologo -W3 -EHsc -DENG_FORMAT_MICRO_GLYPH=\"u\" -I../src test_eng_format.cpp ../src/eng_format.cpp && test_eng_format
// g++ -Wall -Wextra -std=c++11 -Wno-missing-braces -DENG_FORMAT_MICRO_GLYPH=\"u\" -I../src -o test_eng_format.exe test_eng_format.cpp ../src/eng_format.cpp && test_eng_format


src/eng_format.hpp
// Copyright (C) 2005-2009 by Jukka Korpela
// Copyright (C) 2009-2013 by David Hoerl
// Copyright (C) 2013 by Martin Moene
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#ifndef ENG_FORMAT_H_INCLUDED
#define ENG_FORMAT_H_INCLUDED

#include <string>

/**
 * convert a double to the specified number of digits in SI (prefix) or
 * exponential notation, optionally followed by a unit.
 */
std::string
to_engineering_string( double value, int digits, bool exponential, std::string unit = "", std::string separator = " " );

/**
 * convert the output of to_engineering_string() into a double.
 */
double
from_engineering_string( std::string text );

/**
 * step a value by the smallest possible increment.
 */
std::string
step_engineering_string( std::string text, int digits, bool exponential, bool increment );

//
// Extended interface:
//

/**
 * \var eng_prefixed
 * \brief select SI (prefix) presentation: to_engineering_string(), step_engineering_string().
 */

/**
 * \var eng_exponential
 * \brief select exponential presentation: to_engineering_string(), step_engineering_string().
 */

extern struct eng_prefixed_t {} eng_prefixed;
extern struct eng_exponential_t {} eng_exponential;

/**
 * \var eng_increment
 * \brief let step_engineering_string() make a postive step.
 */

/**
 * \var eng_decrement
 * \brief let step_engineering_string() make a negative step.
 */

const bool eng_increment = true;
const bool eng_decrement = false;

/**
 * convert a double to the specified number of digits in SI (prefix) notation,
 * optionally followed by a unit.
 */
inline std::string
to_engineering_string( double value, int digits, eng_prefixed_t, std::string unit = "", std::string separator = " " )
{
    return to_engineering_string( value, digits, false, unit, separator );
}

/**
 * convert a double to the specified number of digits in exponential notation,
 * optionally followed by a unit.
 */
inline std::string
to_engineering_string( double value, int digits, eng_exponential_t, std::string unit = "", std::string separator = " " )
{
    return to_engineering_string( value, digits, true, unit, separator );
}

/**
 * step a value by the smallest possible increment, using SI notation.
 */
inline std::string
step_engineering_string( std::string text, int digits, eng_prefixed_t, bool increment )
{
    return step_engineering_string( text, digits, false, increment );
}

/**
 * step a value by the smallest possible increment, using exponential notation.
 */
inline std::string
step_engineering_string( std::string text, int digits, eng_exponential_t, bool increment )
{
    return step_engineering_string( text, digits, true, increment );
}

#endif // ENG_FORMAT_H_INCLUDED


examples/demo_factors.cpp
// Copyright (C) 2013 by Martin Moene
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "eng_format.hpp"
#include <iostream>
#include <math.h>

int main()
{
    for ( int i = -27; i <= 27; ++i )
    {
        const double x = pow( 10.0, i );
        std::cout << x << " '" << to_engineering_string( x, 3, eng_prefixed   , "V" ) << "'" << std::endl;
        std::cout << x << " '" << to_engineering_string( x, 3, eng_exponential, "V" ) << "'"<< std::endl;
    }

    return 0; // VC6
}

// cl -nologo -W3 -EHsc -I../src demo_factors.cpp ../src/eng_format.cpp && demo_factors.exe
// g++ -Wall -I../src -o demo_factors.exe demo_factors.cpp ../src/eng_format.cpp && demo_factors.exe


examples/demo_eng_format.cpp
// Copyright (C) 2013 by Martin Moene
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "eng_format.hpp"

#include <iomanip>
#include <iostream>
#include <limits>

#include <math.h>

#if defined _MSC_VER && _MSC_VER <= 1200 // VC6
namespace std {
   double max( double const a, double const b ) { return a >= b ? a : b; }
}
#endif

bool approx( double const a, double const b )
{
    const double scale = 1.0;
    const double epsilon = 100 * std::numeric_limits<double>::epsilon();

    return fabs(a - b) < epsilon * (scale + (std::max)( fabs(a), fabs(b) ) );
}

int main()
{
    std::cout <<
       "to_engineering_string( 1e-23, 3, eng_prefixed ): '" <<
        to_engineering_string( 1e-23, 3, eng_prefixed ) << "'\n" <<

       "to_engineering_string( 1234, 3, eng_prefixed, \"Pa\" ): '" <<
        to_engineering_string( 1234, 3, eng_prefixed,  "Pa"  ) << "'\n" <<

       "to_engineering_string( 1234, 3, eng_prefixed, \"Pa\", "" ): '" <<
        to_engineering_string( 1234, 3, eng_prefixed,  "Pa" , "" ) << "'\n" <<

       "step_engineering_string(\"1.00 k\", 3, eng_prefixed, eng_increment ): '" <<
        step_engineering_string( "1.00 k" , 3, eng_prefixed, eng_increment ) << "'\n" << std::endl;

    for ( int i = -24; i <= +24; ++i )
    {
        const double x = +1.00 * pow(10.0,i);

        const std::string text_inp = to_engineering_string( x, 3, eng_prefixed );
        const double             y = from_engineering_string( text_inp );
        const std::string text_out = to_engineering_string( y, 3, eng_prefixed );

        const std::string result = approx( x, y) ? "OK":"ERROR";

        std::cout <<
            result << "\t" <<
            x      << "       \t" <<
            y      << "       \t" <<
            "'"    << text_inp  << "'     \t" <<
            "'"    << text_out  << "'" << std::endl;
    }

    return 0; // VC6
}

// VC6, VC8, VC2010, g++4.8.1

// cl -nologo -W3 -EHsc -DENG_FORMAT_MICRO_GLYPH=\"u\" -I../src demo_eng_format.cpp ../src/eng_format.cpp && demo_eng_format
// g++ -Wall -std=c++03 -DENG_FORMAT_MICRO_GLYPH=\"u\" -I../src -o demo_eng_format.exe demo_eng_format.cpp ../src/eng_format.cpp && demo_eng_format
// g++ -Wall -std=c++11 -DENG_FORMAT_MICRO_GLYPH=\"u\" -I../src -o demo_eng_format.exe demo_eng_format.cpp ../src/eng_format.cpp && demo_eng_format
